components/kyma-environment-broker/internal/process/deprovisioning/remove_service_instance_test.go

```diff
@@ -0,0 +1,264 @@
+
+		// when
+		err = k8sCli.Get(context.TODO(), client.ObjectKey{
+			Namespace: serviceInstanceNamespace,
+			Name:      serviceInstanceName,
+		}, emptySI)
+
+		// then
+		assert.Error(t, err)
+		assert.True(t, k8serrors.IsNotFound(err))
+	})
+
+	t.Run("should not find uaa-issuer service instance and set flag to true", func(t *testing.T) {
+		// given
+		log := logrus.New()
+		ms := storage.NewMemoryStorage()
+
+		scheme := runtime.NewScheme()
+		err := apiextensionsv1.AddToScheme(scheme)
+		decoder := serializer.NewCodecFactory(scheme).UniversalDeserializer()
+		obj, gvk, err := decoder.Decode(siCRD, nil, nil)
+		fmt.Println(gvk)
+		require.NoError(t, err)
+
+		k8sCli := fake.NewClientBuilder().WithScheme(scheme).WithRuntimeObjects(obj).Build()
+		require.NoError(t, err)
+
+		op := fixture.FixDeprovisioningOperation(fixOperationID, fixInstanceID)
+		op.State = "in progress"
+		op.K8sClient = k8sCli
+
+		step := NewRemoveServiceInstanceStep(ms.Operations())
+
+		// when
+		entry := log.WithFields(logrus.Fields{"step": "TEST"})
+		op, _, err = step.Run(op, entry)
+
+		// then
+		assert.NoError(t, err)
+		assert.True(t, op.IsServiceInstanceDeleted)
+	})
+
+	t.Run("should return step repeat time with 20 sec", func(t *testing.T) {
+		// given
+		log := logrus.New()
+		ms := storage.NewMemoryStorage()
+		si := &unstructured.Unstructured{Object: map[string]interface{}{
+			"apiVersion": "services.cloud.sap.com/v1",
+			"kind":       "ServiceInstance",
+			"metadata": map[string]interface{}{
+				"name":      "uaa-issuer",
+				"namespace": "kyma-system",
+			},
+		}}
+
+		k8sCli := fakeK8sClient{}
+		err := k8sCli.Create(context.TODO(), si)
+		require.NoError(t, err)
+
+		op := fixture.FixDeprovisioningOperation(fixOperationID, fixInstanceID)
+		op.State = "in progress"
+		op.K8sClient = &k8sCli
+
+		step := NewRemoveServiceInstanceStep(ms.Operations())
+
+		// when
+		entry := log.WithFields(logrus.Fields{"step": "TEST"})
+		op, repeat, err := step.Run(op, entry)
+
+		// then
+		assert.NoError(t, err)
+		assert.Equal(t, time.Second*20, repeat)
+		assert.False(t, op.IsServiceInstanceDeleted)
+	})
+}
+
+type fakeK8sClient struct {
+	obj *unstructured.Unstructured
+}
+
+func (f *fakeK8sClient) Get(ctx context.Context, key client.ObjectKey, obj client.Object) error {
+	obj = f.obj
+	return nil
+}
+
+func (f *fakeK8sClient) List(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error {
+	return nil
+}
+
+func (f *fakeK8sClient) Create(ctx context.Context, obj client.Object, opts ...client.CreateOption) error {
+	f.obj = new(unstructured.Unstructured)
+	f.obj.SetName(obj.GetName())
+	f.obj.SetNamespace(obj.GetNamespace())
+	return nil
+}
+
+func (f *fakeK8sClient) Delete(ctx context.Context, obj client.Object, opts ...client.DeleteOption) error {
+	return k8serrors.NewInternalError(errors.New("test error"))
+}
+
+func (f *fakeK8sClient) Update(ctx context.Context, obj client.Object, opts ...client.UpdateOption) error {
+	return nil
```

Comment:
nit: for this and below, is it worth returning error making it clearer it's not implemented to not confuse the end user? e.g. `fmt.Errorf("Not implemented in fakeK8sClient")`

Reply:
good point, I will add appropriate messages to avoid confusion in the future